### PR TITLE
#22463 Check for wired ports before wiring in traversal builder

### PR DESCRIPTION
**Pull request by 2m**

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergePreferredSpec.scala [hunk 1]
```diff
@@ -78,6 +78,23 @@ class GraphMergePreferredSpec extends TwoStreamsSetup {
...
+          val merge = b.add(MergePreferred[Int](1))
+
+          s ~> merge.preferred
+          s ~> merge.in(0)
+
+          merge.out ~> Sink.head[Int]
+          merge.out ~> Sink.head[Int]
+          ClosedShape
+        })
+      }).getMessage should include("[MergePreferred.out] is already connected")
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphMergePreferredSpec.scala]: nice message 👍

**Review by ktoso — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 2]
```diff
@@ -996,7 +996,14 @@ object GraphDSL extends GraphApply {
...
     private[GraphDSL] def addEdge[T, U >: T](from: Outlet[T], to: Inlet[U]): Unit =
-      traversalBuilderInProgress = traversalBuilderInProgress.wire(from, to)
+      if (traversalBuilderInProgress.isUnwired(from) && traversalBuilderInProgress.isUnwired(to)) {
+        traversalBuilderInProgress = traversalBuilderInProgress.wire(from, to)
+      } else {
+        if (!traversalBuilderInProgress.isUnwired(from))
+          throw new IllegalArgumentException(s"[${from.s}] is already connected")
+        else if (!traversalBuilderInProgress.isUnwired(to))
+          throw new IllegalArgumentException(s"[${to.s}] is already connected")
+      }
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: would it be worth doing the `wire` optimistically and only in catch exception do the isUnwired checks? some of them are Map lookups.

Comment (2m) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: We would save up a couple of map lookups that way, so it is probably worth it.

**Review by johanandren — APPROVED** [review]
LGTM!

**Comment by drewhk** [comment]
LGTM

**Pull request merged**